### PR TITLE
Q lemmas: reduction, EQ from parts, num/denum cancellation

### LINKED ISSUE #14066: Proposal to add lemmas to Q
I propose adding the following lemmas to the standard library for Q:

```coq
Require Import Psatz.
Require Import QArith.

Lemma Qdiv_inj : forall (a b : Z) (d : positive),
  a=b <-> (a#d)==(b#d).
Proof.
  intros a b c.
  unfold Qeq. cbn. symmetry.
  apply Z.mul_cancel_r.
  lia.
Qed.

Lemma Qsimp : forall (a b c : Z),
  (0 < a)%Z ->
  (0 < c)%Z ->
  b#(Z.to_pos c) == a*b # Z.to_pos (a*c).
Proof.
  intros a b c hagt0 hcgt0.
  unfold Qeq. cbn. symmetry.
  rewrite Z2Pos.id.
  rewrite Z2Pos.id.
  lia. lia. lia.
Qed.

Lemma Qmulz : forall (a : Z) (b : positive) (c : Z),
  (a#b) * inject_Z c == a*c#b.
Proof.
  intros a b c.
  unfold Qeq. cbn.
  lia.
Qed.

Lemma Qeq_split : forall (a b : Z) (c d : positive),
  a = b -> c = d -> a#c == b#d.
Proof.
  intros a b c d Ha_eq_b Hc_eq_d.
  rewrite Ha_eq_b, Hc_eq_d.
  reflexivity.
Qed.
```

### PR DESCRIPTION
**Pull request by koraa**
**Kind:** feature

Fixes / closes #14066

- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: theories/QArith/QArith_base.v
Comment (jfehrle) [inline theories/QArith/QArith_base.v:820]: "with" or "by"?

File: doc/changelog/10-standard-library/14087-q-lemmas/qlemmas.rst
Comment (jfehrle) [inline doc/changelog/10-standard-library/14087-q-lemmas/qlemmas.rst:1]: Please run `dev/tools/make-changelog.sh` to create this entry in the standard format.

Also, perhaps you could summarize the changes rather than just listing the new items?  And mention `QArith.v` so readers can easily find your additions.

**Review by jfehrle — CHANGES_REQUESTED** [review]
Only reviewed documentation.

File: theories/QArith/QArith_base.v
Comment (MSoegtropIMC) [inline theories/QArith/QArith_base.v:822]: I would call this Qmult_inject_Z_l, because it is about Q-multiplying a Q with an injected Z. One could have similar lemmas for add e.g. and there is no good reason why the name of a lemma about multiplying Q with Z should be shorter than for adding. If you want a shorter name you could leave away the inject since it is obvious than a Qmult with a Z requires an inject_Z. I would check how other lemmas with Q and inject_Z are named.

Comment (koraa) [reply theories/QArith/QArith_base.v:822]: Just to confirm; the suggestion is: "Perform the following renames":

```
Qinject_Z_l -> Qmult_inject_Z_l,
Qinject_Z_r -> Qmult_inject_Z_r
```

Anything else?

I don't have any real opinion on that, so I am perfectly happy to make that change! Do you judge that suggestion to be an obvious improvement or is this something we should wait for someone to weigh in?

Comment (MSoegtropIMC) [reply theories/QArith/QArith_base.v:832]: Yes, I think these names would be good choices.

I would say Qmult_inject_Z_l is a better name because it is a lemma about Qmult. Most lemmas with Qmult as top operator start with Qmult as the following query shows:
```
Require Import QArith.
Search (_*_)%Q.
```
(the lemmas returned by this which do not have Qmult in their name do not have Qmult as top operator). When I add lemmas to the standard library I use such queries to judge what might be a good name.

Otherwise I am fine with the naming.

Thanks for your patience with my nit-picking!

File: doc/changelog/10-standard-library/14087-q-lemmas/qlemmas.rst
Comment (jfehrle) [inline doc/changelog/10-standard-library/14087-q-lemmas/qlemmas.rst:5]: Generally there is only one changelog entry for each PR (not 3), usually with a single sentence summary.  See `doc/changelog` for examples.  The changelog is not meant to be documentation.

Also, it's probably enough just to mention `ZArith` rather than give pathnames.

**Review by jfehrle — CHANGES_REQUESTED** [review]
Probably also worth squashing down to a single commit before merging.  Separation into multiple commits is probably not valuable for the long term.

File: doc/changelog/10-standard-library/14087-q-lemmas/qlemmas.rst
Comment (koraa) [reply doc/changelog/10-standard-library/14087-q-lemmas/qlemmas.rst]: I've squashed and shortened the changelog entry.

If there are more small-ish changes like these – and there may well be some as I am obviously not yet acquainted with coding practices in the coq std lib – I would suggest that you just apply them while merging as this would probably faster for both of us compared to this back and forth.
If you do that please notify me so I can compare and see how to improve my PR next time!

Thanks for the timely reviews :pray:

Comment (jfehrle) [reply doc/changelog/10-standard-library/14087-q-lemmas/qlemmas.rst]: It looks good to me now, thanks.

Generally reviewers don't modify PRs for IMHO a couple reasons:
- Reviewers are a scarce resource (as are submitters!).  All reviewers are volunteers who have many demands on their time (even the Inria researchers who mostly work on Coq).
- Extra overhead to pull the PR code onto the reviewer's system and later push to GitHub again.  If a reviewer does that and the submitter makes subsequent changes, the submitter needs to pull the reviewer's changes.  This is prone to user errors that lose changes (i.e. forgetting to pull such changes).
- We want all changes to have independent review.  If the first reviewer makes significant changes, then we expect another reviewer will give the final approval.

Specific reviewers are expected based on what code is changed.  I mostly review documentation.  We need an "asignee" to give final approval and merge the code.  @herbelin, are you available to do that?  (or @Zimmi48 do you have another suggestion?)

Comment (jfehrle) [reply doc/changelog/10-standard-library/14087-q-lemmas/qlemmas.rst]: PS I unresolved the conversation so it's not hidden.

**Comment by herbelin** [comment]
>  We need an "assignee" to give final approval and merge the code. @herbelin, are you available to do that? 

I don't know so well the QArith library. @MSoegtropIMC is certainly of much better advice, if he does not mind to assign.

**Comment by MSoegtropIMC** [comment]
@herbelin : Sure I can take care of it. Btw.: should I join the standard library maintainers group?

**Comment by Zimmi48** [comment]
> @herbelin : Sure I can take care of it. Btw.: should I join the standard library maintainers group?

Sure, feel free to do that if you wish.

File: theories/QArith/QArith_base.v
Comment (silene) [inline theories/QArith/QArith_base.v:68]: I am not sure this lemma is worth it. As shown by the proof, it is just two trivial rewriting away.

File: theories/QArith/QArith_base.v
Comment (silene) [inline theories/QArith/QArith_base.v:76]: Using `<->` makes things more painful in general. If the lemma had just been `(a#p)==(b#p) -> a=b`, it could for example be used as a rewriting rule. (The `<-` direction is not that interesting anyway.)

File: theories/QArith/QArith_base.v
Comment (silene) [inline theories/QArith/QArith_base.v:84]: There is no reason for `Qnum_cancel` to be oriented the opposite way with respect to `Qden_cancel`.

And as with `Qden_cancel`, the only interesting direction is `(z#a)==(z#b) -> a = b`.

File: theories/QArith/QArith_base.v
Comment (silene) [inline theories/QArith/QArith_base.v:89]: There is no reason to perform an induction over `z`. This lemma is just a corollary of `Zmult_reg_l`.

File: theories/QArith/QArith_base.v
Comment (silene) [inline theories/QArith/QArith_base.v:828]:
```suggestion
  Construction of a new rational by multiplication with an integer
```

File: doc/changelog/10-standard-library/14087-q-lemmas/qlemmas.rst
Comment (Zimmi48) [reply doc/changelog/10-standard-library/14087-q-lemmas/qlemmas.rst]: @jfehrle Sometimes it is indeed more efficient for reviewers to push changes directly (avoids a lot of back and forth). If the changes are trivial, this is fine. If not, then the reviewer becomes a co-author and cannot be the assignee. GitHub tooling can help to make checking out a PR locally a breeze (see the "open with" button at the top-left of the PR). Pushing is a bit more complex: you need to use `git push FORK_URL`. And it is true that the submitter needs to remember pulling the reviewer's changes (the same is true with the use of inline suggestions during reviews though).

File: theories/QArith/QArith_base.v
Comment (koraa) [reply theories/QArith/QArith_base.v:76]: It is my understanding that `<->` can be used for rewriting bidirectionally but `->` can not in either direction?

```coq
Goal forall (a b : Z) (p : positive) (q u : Q),
  (a#p) == (b#p) /\ q == u <-> a=b /\ q == u.
Proof.
  intros a b p q u.
  rewrite (Qden_cancel a b p).
  reflexivity.
Qed.

Goal forall (a b : Z) (p : positive) (q u : Q),
  (a#p) == (b#p) /\ q == u <-> a=b /\ q == u.
Proof.
  intros a b p q u.
  rewrite <- (Qden_cancel a b p).
  reflexivity.
Qed.
```

File: theories/QArith/QArith_base.v
Comment (koraa) [reply theories/QArith/QArith_base.v:84]: Oh right that is indeed an oversight; I will flip the order!

**Comment by Zimmi48** [comment]
@jfehrle your review is still marked as a change request

File: theories/QArith/QArith_base.v
Comment (koraa) [reply theories/QArith/QArith_base.v:68]: Most proofs in QArith_base are this simple; some most involve ring or apply or auto at some point but to the user of the lemmas I'd say this doesn't make much of a difference.

I think providing it is worth it as the intended use case is apply, thereby introducing goals for den/num equality and there is didactic worth in that users can find this as a viable proof strategy for establishing equality (instead of unfold of ==). Similar to split. But of course, using assert is always an option…so I am happy to delete if you disagree.

File: theories/QArith/QArith_base.v
Comment (koraa) [reply theories/QArith/QArith_base.v:828]: Thanks

File: theories/QArith/QArith_base.v
Comment (silene) [reply theories/QArith/QArith_base.v:76]: We are talking about two separate things. When rewriting with your version of `Qden_cancel`, the reflexive, symmetric, transitive relation is neither `=` nor `==`, it is `<->`. In other words, your goal needs to contain a complete equality, so that it can be replaced by another complete equality. This almost never happens in practice.

What often happens, however, is that you have obtained a hypothesis `a#p == b#p` and you thus want to replace all the occurrences of `a` with `b` in your goal. This is possible with your version, but quite painful. That is why I am suggesting to change the lemma to the following:
```coq
Lemma Qden_cancel : forall (a b : Z) (p : positive),
  (a#p)==(b#p) -> a=b.
Admitted.

Goal forall (a b : Z) (p : positive),
  (a#p) == (b#p) -> (a - b = 0)%Z.
Proof.
  intros a b p H.
  rewrite Qden_cancel with (1 := H).
  ring.
Qed.
```

Comment (koraa) [reply theories/QArith/QArith_base.v:76]: Maybe I am missing something here? I though it should be way easier to derive A -> B, B -> A and B <-> A from A <-> B than juggling all these _l _r versions? In desparate cases, if I need to apply a lemma, by using iff_sym…

```coq
Lemma Qden_cancel_rev : forall (a b : Z) (p : positive),
  a=b <-> (a#p)==(b#p).
Proof.
  intros a b p.
  exact (iff_sym (Qden_cancel _ _ _)).
Qed.
```

I though there must be some better/prettier way to achieve this yet and I just missed it because I skimmed the doc instead of reading it…

File: theories/QArith/QArith_base.v
Comment (silene) [reply theories/QArith/QArith_base.v:68]: That is where I tend to put the bar: Is it faster for me to prove a property on the fly or to search for its name in the standard library? Here, finding and typing a sufficiently specific pattern for `Search` then perusing through all the results and finally applying the found lemma would take me much more time than just doing `rewrite` or `replace` twice. That is why I tend to classify it in the category of useless lemmas. Your mileage may vary.

File: theories/QArith/QArith_base.v
Comment (silene) [reply theories/QArith/QArith_base.v:77]: You can derive them by using the first and second projection (`proj1` and `proj2`), since `<->` is just a conjunction. But you can apply these projections only if you have a fully applied term. For `Qden_cancel`, it is a tedious but straightforward. But for `Qnum_cancel`, this becomes very painful, especially if you are trying to use it as a rewriting rule.

```coq
Lemma Qnum_cancel : forall (a b : positive) (z : Z),
  z<>0%Z -> (z#a)==(z#b) <-> a=b.
Admitted.

Goal forall a p q, (a > 0)%Z ->
  (a#p) == (a#q) + 0 -> (Zpos p - Zpos q = 0)%Z.
Proof.
  intros a p q H H'.
  erewrite (proj1 (Qnum_cancel p q a _)).
  ring.
  rewrite H'.
  ring.
  Unshelve.
  lia.
Qed.
```

If `Qnum_cancel` had been an implication as I am suggesting, you would have been able to simply type `rewrite (Qnum_cancel p q a).` No need to know about `erewrite`, `proj1`, `Unshelve`. (Obviously, you can avoid all of those by adding sufficiently many `assert` to your proof, but that is still tedious.)

Comment (koraa) [reply theories/QArith/QArith_base.v:77]: Convinced, thank you for the explanation!

**Review by jfehrle — APPROVED** [review]

**Comment by MSoegtropIMC** [comment]
Just wanted to apologize for the delay - I am busy until Thursday. I will do the final review and merge then.

**Comment by koraa** [comment]
> Just wanted to apologize for the delay - I am busy until Thursday. I will do the final review and merge then.

That is perfectly all right; figured something like this. Thank you!

**Review by MSoegtropIMC — APPROVED** [review]
LGTM - will merge now.

**Comment by MSoegtropIMC** [comment]
@koraa : congratulations for your first merged Coq PR :-)

**Comment by koraa** [comment]
Thank you and in fact I am a bit too proud than would be appropriate for lemmas of that magnitude :D

**Comment by MSoegtropIMC** [comment]
I hope that was just the beginning!

Please note that I am currently preparing a larger update on Q, mostly moving lemmas from `theories/Reals/Cauchy/QExtra.v` to where they belong. It should be finished on Monday. So if you plan to add more, please have a look there if it is not already there.

**Comment by MSoegtropIMC** [comment]
@koraa : the PR I mentioned above is #14293.

**Pull request merged**
